outlines/fsm/guide.py

```diff
@@ -193,12 +193,8 @@ def get_next_state(self, state: int, token_id: int) -> int:
         The new state of the guide.
 
         """
-        if token_id == self.eos_token_id:
+        if token_id == self.eos_token_id or state not in self.states_to_token_maps:
```

Comment:
@br3no  I was wondering if we really need the 2nd condition `state not in self.states_to_token_maps` ? The condition basically checks for states which do not have outgoing edges which would be a part of final states in the FSM. [This block of code](https://github.com/outlines-dev/outlines/blob/7863f8e8bbaeb71c9d2434636a2d63bfe6dd7d39/outlines/fsm/regex.py#L835-L841) add EOS as an edge to such states which makes them have atleast one outgoing edges and no states in FSM will be absent in the `states_to_token_maps`. Wdyt?

Reply:
@ekagra-ranjan yes, we do need this second condition.

The block of code you linked to does not add an EOS outbound transition to these states. It only adds transitions to final states which are present in `states_to_token_maps`. But these states are not present there. `states_to_token_subsets.get(state)` will return `None` for these states.

I'm not really knowledgeable about the way Outlines (and interegular) build the state machines out of regexes. The matter of fact is that the `states_to_token_maps` does **not** contain all states that are reachable. I have noticed this while debugging the code for some example regexes. 

This is not a problem in principle, as these states are considered to be final and `states_to_token_subsets.get(state) is None` is used all over the code to handle this special case (as in the block you linked to). 

I actually believe this could be improved and Outlines would profit from removing this special case that needs to be thought of all over the place and could lead to bugs. But this is, as I said, not a problem in principle.